lib/libnvpair/libnvpair.c

```diff
@@ -199,6 +199,11 @@ nvprint_##type_and_variant(nvlist_prtctl_t pctl, void *private, \
 	return (1); \
 }
 
+#if defined(__GNUC__) && !defined(__clang__) && \
+    defined(ZFS_GCC_FORMAT_OVERFLOW_SUPPORTED)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-overflow"
```

Comment:
I am curious what warnings the compiler emits without this. Perhaps, it could be handled in a way that does not disable the check.

Reply:
```
  CC       lib/libnvpair/libnvpair_la-libnvpair.lo
In file included from /usr/include/stdio.h:906,
                 from ./include/libnvpair.h:31,
                 from lib/libnvpair/libnvpair.c:32:
In function 'fprintf',
    inlined from 'nvprint_boolean' at lib/libnvpair/libnvpair.c:207:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_boolean_value' at lib/libnvpair/libnvpair.c:208:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_byte' at lib/libnvpair/libnvpair.c:209:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_int8' at lib/libnvpair/libnvpair.c:210:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_uint8' at lib/libnvpair/libnvpair.c:211:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_int16' at lib/libnvpair/libnvpair.c:212:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_uint16' at lib/libnvpair/libnvpair.c:213:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_int32' at lib/libnvpair/libnvpair.c:214:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_uint32' at lib/libnvpair/libnvpair.c:215:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_int64' at lib/libnvpair/libnvpair.c:216:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_uint64' at lib/libnvpair/libnvpair.c:217:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_double' at lib/libnvpair/libnvpair.c:218:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_string' at lib/libnvpair/libnvpair.c:219:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
In function 'fprintf',
    inlined from 'nvprint_hrtime' at lib/libnvpair/libnvpair.c:220:1:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
   79 |   return __fprintf_chk (__stream, __USE_FORTIFY_LEVEL - 1, __fmt,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   80 |                         __va_arg_pack ());
      |                         ~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/bits/stdio2.h:79:10: error: null format string [-Werror=format-overflow=]
```

Reply:
Is it Clang or GCC complaining about this and is this complaint only in userspace?

I am using both GCC 12.2.0 and LLVM/Clang 15.0.2 on Gentoo and I am not hitting this issue. Presumably, Canonical would have encountered this too since they build ZFS on Ubuntu:

https://www.ubuntuupdates.org/package/core/kinetic/main/base/zfsutils-linux

I am also using glibc 2.36 on Gentoo, which is what Ubuntu 22.10 uses:

https://www.ubuntuupdates.org/package/core/kinetic/main/base/glibc

I am curious why I am not hitting this locally, despite the package versions seemingly matching.

Reply:
I just run a compilation in a Ubuntu 22.10 VM on my Mac (`arm64`) and none of those bugs could be reproduced! Seems like the issue is specific to `amd64`.
Bugs pop out only in userspace. The kernel module has been compiled correctly.

Reply:
I cannot reproduce this on a Ubuntu 22.10 userland on amd64:

https://github.com/openzfs/zfs/pull/14150#issuecomment-1309849929